patroni/postgresql/__init__.py

```diff
@@ -984,16 +988,31 @@ def pick_synchronous_standby(self, cluster, sync_node_count=1):
         # receiving changes faster than the sync member (very rare but possible). Such cases would
         # trigger sync standby member swapping frequently and the sort on sync_state desc should
         # help in keeping the query result consistent.
-        for app_name, state, sync_state in self.query(
-                "SELECT pg_catalog.lower(application_name), state, sync_state"
+        current_wal_lsn = int(str(self.last_operation()))
+        for app_name, state, sync_state, replica_lag in self.query(
+                "SELECT app_name, state, sync_state,"
+                "       (case when replica_count > 0 then max_replica_lsn - replica_lsn"
+                "             else {3} - replica_lsn"
+                "        end) replica_lag"
+                " FROM ("
+                " SELECT pg_catalog.lower(application_name) app_name, state, sync_state"
+                "       , pg_{2}_{1}_diff({0}_{1}, '0/0')::bigint replica_lsn"
+                "       , max( pg_{2}_{1}_diff({0}_{1}, '0/0')::bigint ) over () max_replica_lsn"
+                "       , count(1) over () replica_count"
```

Comment:
The problem is that not all connected replicas are real cluster members.
It could be for example pg_receivewal, or standby cluster and they should not be used in the `MAX()` and `COUNT()`.

We have two options to handle it:
1. Filter by application_name in the query
2. Fetch the whole result set to python and process it.

I would prefer the second one.

Reply:
Thanks. Fixed the code.